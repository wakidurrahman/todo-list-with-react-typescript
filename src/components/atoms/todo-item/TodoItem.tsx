/**
 * TodoItem Component
 *
 * A React component that displays a single todo item in a card format with the following features:
 * - Shows todo title, description, and last updated timestamp.
 * - Checkbox to toggle completion status.
 * - Edit and Delete action buttons.
 * - Responsive layout with Bootstrap.
 * - Accessibility features.
 * - Visual indication of completed todos.
 *
 * @component
 * @example
 * ```tsx
 * <TodoItem
 *   todo={todoItem}
 *   onEdit={handleEdit}
 *   onDelete={handleDelete}
 *   onToggleComplete={handleToggleComplete}
 * />
 * ```
 *
 * @param {Object} props - Component props
 * @param {Todo} props.todo - The todo item to display
 * @param {Function} props.onEdit - Callback when edit button is clicked
 * @param {Function} props.onDelete - Callback when delete button is clicked
 * @param {Function} props.onToggleComplete - Optional callback when completion status is toggled
 */

import React from 'react';

import Button from '@/components/atoms/button/Button';
import Checkbox from '@/components/atoms/checkbox/Checkbox';
import { Priority, Todo } from '@/types/todo.types';

type TodoItemProps = {
  todo: Todo;
  onEdit: (todo: Todo) => void;
  onDelete: (id: string) => void;
  onToggleComplete?: (id: string) => void;
};

// Add priority colors
const priorityColors: Record<Priority, string> = {
  high: 'danger',
  medium: 'warning',
  low: 'info',
};

const TodoItem: React.FC<TodoItemProps> = ({
  todo,
  onEdit,
  onDelete,
  onToggleComplete,
}) => {
  const isOverdue = todo.dueDate && new Date(todo.dueDate) < new Date();

  return (
    <div className="card mb-1">
      <div className="card-body">
        <div className="row align-items-start g-2">
          {/* Checkbox and Content Column */}
          <div className="col">
            <Checkbox
              checked={todo.completed ?? false}
              onChange={() => onToggleComplete?.(todo.id)}
              id={`todo-${todo.id}`}
              value={todo.id}
            >
              <div className="ms-2">
                <div className="d-flex align-items-center gap-2 mb-3">
                  <h5 className="card-title mb-0 text-break">{todo.title}</h5>
                  <span
                    className={`badge bg-${priorityColors[todo.priority as Priority]}`}
                  >
                    {todo.priority}
                  </span>
                </div>

                <h6 className="card-subtitle text-body-secondary mb-2">
                  <i className="bi bi-clock me-1"></i>
                  Updated:{' '}
                  {todo.updatedAt
                    ? new Date(todo.updatedAt).toLocaleString()
                    : 'Never'}
                </h6>
                {todo.dueDate && (
                  <h6
                    className={`card-subtitle mb-0 ${isOverdue ? 'text-danger' : 'text-body-secondary'}`}
                  >
                    <i className="bi bi-calendar-event me-1"></i>
                    Due: {new Date(todo.dueDate).toLocaleString()}
                  </h6>
                )}

                {todo.description && (
                  <p className="card-text text-break mt-2 mb-0">
                    {todo.description}
                  </p>
                )}
              </div>
            </Checkbox>
          </div>

          {/* Action Buttons Column */}
          <div className="col-auto">
            <div className="btn-group">
              <Button
                variant="secondary"
                size="sm"
                onClick={() => onEdit(todo)}
                aria-label="Edit todo"
                data-testid={`edit-todo-${todo.id}`}
              >
                <i className="bi bi-pencil-square me-1"></i>
              </Button>
              <Button
                variant="danger"
                size="sm"
                onClick={() => onDelete(todo.id)}
                aria-label="Delete todo"
                data-testid={`delete-todo-${todo.id}`}
              >
                <i className="bi bi-trash me-1"></i>
              </Button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default TodoItem;
