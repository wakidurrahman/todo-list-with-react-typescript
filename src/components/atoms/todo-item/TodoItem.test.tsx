import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TodoItem from './TodoItem';

describe('TodoItem Component', () => {
  const mockTodo = {
    id: '1',
    title: 'Test Todo',
    description: 'Test Description',
    createdAt: new Date('2024-02-16T09:00:00.000Z'),
    updatedAt: new Date('2024-02-16T09:00:00.000Z'),
  };

  // Snapshot Tests
  describe('Snapshots', () => {
    it('renders todo item correctly', () => {
      const { container } = render(
        <TodoItem todo={mockTodo} onEdit={() => {}} onDelete={() => {}} />
      );
      expect(container).toMatchSnapshot();
    });
  });

  // Unit Tests
  describe('Functionality', () => {
    const user = userEvent.setup();

    it('displays todo title and description', () => {
      render(
        <TodoItem todo={mockTodo} onEdit={() => {}} onDelete={() => {}} />
      );

      expect(screen.getByText(mockTodo.title)).toBeInTheDocument();
      expect(screen.getByText(mockTodo.description)).toBeInTheDocument();
    });

    it('calls onEdit when edit button is clicked', async () => {
      const handleEdit = jest.fn();
      render(
        <TodoItem todo={mockTodo} onEdit={handleEdit} onDelete={() => {}} />
      );

      const editButton = screen.getByRole('button', { name: /edit/i });
      await user.click(editButton);

      expect(handleEdit).toHaveBeenCalledWith(mockTodo);
    });

    it('calls onDelete when delete button is clicked', async () => {
      const handleDelete = jest.fn();
      render(
        <TodoItem todo={mockTodo} onEdit={() => {}} onDelete={handleDelete} />
      );

      const deleteButton = screen.getByRole('button', { name: /delete/i });
      await user.click(deleteButton);

      expect(handleDelete).toHaveBeenCalledWith(mockTodo.id);
    });

    it('displays formatted dates', () => {
      render(
        <TodoItem todo={mockTodo} onEdit={() => {}} onDelete={() => {}} />
      );

      // Using a partial match for the date since formatting may vary
      const datePattern = /2\/16\/2024/;
      expect(screen.getByText(datePattern)).toBeInTheDocument();
    });

    it('has buttons with proper accessibility features', () => {
      render(
        <TodoItem todo={mockTodo} onEdit={() => {}} onDelete={() => {}} />
      );

      const editButton = screen.getByRole('button', { name: /edit/i });
      const deleteButton = screen.getByRole('button', { name: /delete/i });

      expect(editButton.querySelector('.bi-pencil-square')).toBeInTheDocument();
      expect(editButton).toHaveTextContent(/edit/i);

      expect(deleteButton.querySelector('.bi-trash')).toBeInTheDocument();
      expect(deleteButton).toHaveTextContent(/delete/i);
    });
  });
});
