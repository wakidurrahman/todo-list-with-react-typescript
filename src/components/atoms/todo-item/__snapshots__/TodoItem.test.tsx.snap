// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem Component Snapshots matches basic todo snapshot 1`] = `
<div
  className="card mb-1"
  data-testid="todo-item-1"
>
  <div
    className="card-body"
  >
    <div
      className="row align-items-start g-2"
    >
      <div
        className="col"
      >
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input mt-2"
            id="todo-1"
            onChange={[Function]}
            type="checkbox"
            value="1"
          />
          <label
            className="form-check-label w-100 "
            htmlFor="todo-1"
          >
            <div
              className="ms-2"
            >
              <div
                className="d-flex align-items-center gap-2 mb-3"
              >
                <h5
                  className="card-title mb-0 text-break "
                >
                  Test Todo
                </h5>
                <span
                  className="badge bg-danger"
                  title="Priority: high"
                >
                  high
                </span>
              </div>
              <h6
                className="card-subtitle text-body-secondary mb-2"
              >
                <i
                  aria-hidden="true"
                  className="bi bi-clock me-1"
                />
                Updated:
                 
                2/23/2024, 9:00:00 PM
              </h6>
              <h6
                className="card-subtitle mb-0 text-danger"
              >
                <i
                  aria-hidden="true"
                  className="bi bi-calendar-event me-1"
                />
                Due: 
                2/24/2024, 9:00:00 PM
              </h6>
              <p
                className="card-text text-break mt-2 mb-0 "
              >
                Test Description
              </p>
            </div>
          </label>
        </div>
      </div>
      <div
        className="col-auto"
      >
        <div
          className="btn-group"
        >
          <button
            aria-label="Edit todo "Test Todo""
            className="btn btn-secondary btn-sm"
            data-testid="edit-todo-1"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-pencil-square me-1"
            />
            Edit
          </button>
          <button
            aria-label="Delete todo "Test Todo""
            className="btn btn-danger btn-sm"
            data-testid="delete-todo-1"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-trash me-1"
            />
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoItem Component Snapshots matches completed todo snapshot 1`] = `
<div
  className="card mb-1"
  data-testid="todo-item-1"
>
  <div
    className="card-body"
  >
    <div
      className="row align-items-start g-2"
    >
      <div
        className="col"
      >
        <div
          className="form-check"
        >
          <input
            checked={true}
            className="form-check-input mt-2"
            id="todo-1"
            onChange={[Function]}
            type="checkbox"
            value="1"
          />
          <label
            className="form-check-label w-100 text-decoration-line-through"
            htmlFor="todo-1"
          >
            <div
              className="ms-2"
            >
              <div
                className="d-flex align-items-center gap-2 mb-3"
              >
                <h5
                  className="card-title mb-0 text-break text-decoration-line-through"
                >
                  Test Todo
                </h5>
                <span
                  className="badge bg-danger"
                  title="Priority: high"
                >
                  high
                </span>
              </div>
              <h6
                className="card-subtitle text-body-secondary mb-2"
              >
                <i
                  aria-hidden="true"
                  className="bi bi-clock me-1"
                />
                Updated:
                 
                2/23/2024, 9:00:00 PM
              </h6>
              <h6
                className="card-subtitle mb-0 text-danger"
              >
                <i
                  aria-hidden="true"
                  className="bi bi-calendar-event me-1"
                />
                Due: 
                2/24/2024, 9:00:00 PM
              </h6>
              <p
                className="card-text text-break mt-2 mb-0 text-decoration-line-through"
              >
                Test Description
              </p>
            </div>
          </label>
        </div>
      </div>
      <div
        className="col-auto"
      >
        <div
          className="btn-group"
        >
          <button
            aria-label="Edit todo "Test Todo""
            className="btn btn-secondary btn-sm"
            data-testid="edit-todo-1"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-pencil-square me-1"
            />
            Edit
          </button>
          <button
            aria-label="Delete todo "Test Todo""
            className="btn btn-danger btn-sm"
            data-testid="delete-todo-1"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-trash me-1"
            />
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoItem Component Snapshots matches minimal todo snapshot 1`] = `
<div
  className="card mb-1"
  data-testid="todo-item-2"
>
  <div
    className="card-body"
  >
    <div
      className="row align-items-start g-2"
    >
      <div
        className="col"
      >
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input mt-2"
            id="todo-2"
            onChange={[Function]}
            type="checkbox"
            value="2"
          />
          <label
            className="form-check-label w-100 "
            htmlFor="todo-2"
          >
            <div
              className="ms-2"
            >
              <div
                className="d-flex align-items-center gap-2 mb-3"
              >
                <h5
                  className="card-title mb-0 text-break "
                >
                  Minimal Todo
                </h5>
                <span
                  className="badge bg-info"
                  title="Priority: low"
                >
                  low
                </span>
              </div>
              <h6
                className="card-subtitle text-body-secondary mb-2"
              >
                <i
                  aria-hidden="true"
                  className="bi bi-clock me-1"
                />
                Updated:
                 
                Never
              </h6>
            </div>
          </label>
        </div>
      </div>
      <div
        className="col-auto"
      >
        <div
          className="btn-group"
        >
          <button
            aria-label="Edit todo "Minimal Todo""
            className="btn btn-secondary btn-sm"
            data-testid="edit-todo-2"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-pencil-square me-1"
            />
            Edit
          </button>
          <button
            aria-label="Delete todo "Minimal Todo""
            className="btn btn-danger btn-sm"
            data-testid="delete-todo-2"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-trash me-1"
            />
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoItem Component Snapshots matches overdue todo snapshot 1`] = `
<div
  className="card mb-1"
  data-testid="todo-item-1"
>
  <div
    className="card-body"
  >
    <div
      className="row align-items-start g-2"
    >
      <div
        className="col"
      >
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input mt-2"
            id="todo-1"
            onChange={[Function]}
            type="checkbox"
            value="1"
          />
          <label
            className="form-check-label w-100 "
            htmlFor="todo-1"
          >
            <div
              className="ms-2"
            >
              <div
                className="d-flex align-items-center gap-2 mb-3"
              >
                <h5
                  className="card-title mb-0 text-break "
                >
                  Test Todo
                </h5>
                <span
                  className="badge bg-danger"
                  title="Priority: high"
                >
                  high
                </span>
              </div>
              <h6
                className="card-subtitle text-body-secondary mb-2"
              >
                <i
                  aria-hidden="true"
                  className="bi bi-clock me-1"
                />
                Updated:
                 
                2/23/2024, 9:00:00 PM
              </h6>
              <h6
                className="card-subtitle mb-0 text-danger"
              >
                <i
                  aria-hidden="true"
                  className="bi bi-calendar-event me-1"
                />
                Due: 
                2/22/2025, 10:23:00 PM
              </h6>
              <p
                className="card-text text-break mt-2 mb-0 "
              >
                Test Description
              </p>
            </div>
          </label>
        </div>
      </div>
      <div
        className="col-auto"
      >
        <div
          className="btn-group"
        >
          <button
            aria-label="Edit todo "Test Todo""
            className="btn btn-secondary btn-sm"
            data-testid="edit-todo-1"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-pencil-square me-1"
            />
            Edit
          </button>
          <button
            aria-label="Delete todo "Test Todo""
            className="btn btn-danger btn-sm"
            data-testid="delete-todo-1"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <i
              aria-hidden="true"
              className="bi bi-trash me-1"
            />
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TodoItem Component Snapshots renders todo item correctly 1`] = `
<div>
  <div
    class="card mb-1"
    data-testid="todo-item-1"
  >
    <div
      class="card-body"
    >
      <div
        class="row align-items-start g-2"
      >
        <div
          class="col"
        >
          <div
            class="form-check"
          >
            <input
              class="form-check-input mt-2"
              id="todo-1"
              type="checkbox"
              value="1"
            />
            <label
              class="form-check-label w-100 "
              for="todo-1"
            >
              <div
                class="ms-2"
              >
                <div
                  class="d-flex align-items-center gap-2 mb-3"
                >
                  <h5
                    class="card-title mb-0 text-break "
                  >
                    Test Todo
                  </h5>
                  <span
                    class="badge bg-undefined"
                    title="Priority: undefined"
                  />
                </div>
                <h6
                  class="card-subtitle text-body-secondary mb-2"
                >
                  <i
                    aria-hidden="true"
                    class="bi bi-clock me-1"
                  />
                  Updated:
                   
                  2/16/2024, 6:00:00 PM
                </h6>
                <p
                  class="card-text text-break mt-2 mb-0 "
                >
                  Test Description
                </p>
              </div>
            </label>
          </div>
        </div>
        <div
          class="col-auto"
        >
          <div
            class="btn-group"
          >
            <button
              aria-label="Edit todo "Test Todo""
              class="btn btn-secondary btn-sm"
              data-testid="edit-todo-1"
              type="button"
            >
              <i
                aria-hidden="true"
                class="bi bi-pencil-square me-1"
              />
              Edit
            </button>
            <button
              aria-label="Delete todo "Test Todo""
              class="btn btn-danger btn-sm"
              data-testid="delete-todo-1"
              type="button"
            >
              <i
                aria-hidden="true"
                class="bi bi-trash me-1"
              />
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
